misc/makefile_help.go

```diff
@@ -0,0 +1,381 @@
+	line = strings.TrimSpace(strings.TrimLeft(line, "# "))
+	line = prefixRe.ReplaceAllString(line, "")
+	if line == "" {
+		return "", nil
+	}
+	return fmt.Sprintf(" (%s)", line), nil
+}
+
+// scrapeReadmeBanners returns a map[dir]banner for each dir in wildcards or dirs.
+func scrapeReadmeBanners(wildcards, dirs []string) map[string]string {
+	banners := make(map[string]string)
+	for _, group := range [][]string{wildcards, dirs} {
+		for _, d := range group {
+			if _, seen := banners[d]; !seen {
+				b, _ := readReadmeBanner(d)
+				banners[d] = b
+			}
+		}
+	}
+	return banners
+}
+
+// printTargets lists all targets and descriptions, expanding '%' with wildcards.
+func printTargets(w io.Writer, targets map[string]string, wildcards [][]string, banners map[string]string) {
+	keys := make([]string, 0, len(targets))
+	for k := range targets {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for i := range wildcards {
+		sort.Strings(wildcards[i])
+	}
+
+	width := maxKeyLength(keys, wildcards)
+	for _, key := range keys {
+		desc := targets[key]
+		wildIdx := 0
+
+		if m := percentNumPrefixRE.FindStringSubmatch(desc); m != nil {
+			idx, err := strconv.Atoi(m[1])
+			if err != nil || idx <= 0 {
+				// This should not be reachable.
+				panic(fmt.Sprintf("invalid prefix index in %q", desc))
+			}
+			wildIdx = idx - 1
+			desc = desc[len(m[0]):]
+		}
+
+		if strings.Contains(key, "%") && (len(wildcards) > wildIdx) && (len(wildcards[wildIdx]) > 0) {
+			for _, wild := range wildcards[wildIdx] {
+				ek := strings.ReplaceAll(key, "%", wild)
+				if desc == "" {
+					fmt.Fprintf(w, "  %s\n", ek)
+				} else {
+					ed := strings.ReplaceAll(desc, "%", wild)
+					fmt.Fprintf(w, "  %-*s   <-- %s%s\n", width, ek, ed, banners[wild])
+				}
+			}
+		} else {
+			if desc == "" {
+				fmt.Fprintf(w, "  %s\n", key)
+			} else {
+				fmt.Fprintf(w, "  %-*s   <-- %s\n", width, key, desc)
+			}
+		}
+	}
+}
+
+// printSubdirs scans each dir for a Makefile, marks '*' if it has a 'help' target,
+// and prints a make -C invocation with any README banner.
+func printSubdirs(w io.Writer, relativeTo string, dirs []string, banners map[string]string) {
+	if len(dirs) == 0 {
+		return
+	}
+	fmt.Fprintln(w, "\nSub‑directories with make targets:")
+	sort.Strings(dirs)
+
+	noteHelp := false
+	for _, d := range dirs {
+		mf := filepath.Join(d, "Makefile")
+		info, err := os.Stat(mf)
+		if err != nil || info.IsDir() {
+			continue
+		}
+		tMap, err := extractMakefileTargets(mf)
+		hasHelp := false
+		if err == nil {
+			_, ok := tMap["help"]
+			hasHelp = ok
+		}
+		star := " "
+		if hasHelp {
+			star = "*"
+			noteHelp = true
+		}
+		targetDir := d
+		if relativeTo != "" {
+			targetDir = filepath.ToSlash(filepath.Join(relativeTo, d))
+		}
+		fmt.Fprintf(w, "    %s  make -C %s%s\n", star, targetDir, banners[d])
+	}
+	if noteHelp {
+		fmt.Fprintln(w, "\n       * Is documented with a `help` target.")
+	}
+}
+
+func run(args []string, stdout, stderr io.Writer) int {
```

Comment:
Unless you have a good reason to distinguish between return code 1 and 2, I suggest you make this return an `error` and do error handling in `main` (where if an error was returned, it exits with 1)

Reply:
Just to clarify, I don't care about having multiple exit code values; but I do care about keeping `main()` a thin wrapper without needing error handling.

Reply:
It's trivial and idiomatic to do this in Go:

```go
func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "fatal: %v\n", err)
		os.Exit(1)
	}
}
```

I'd prefer this, but feel free to resolve it.

Reply:
`run()` now returns `error`. New (trivial) function `ErrorToExitCode()` converts that to an exit code.